### PR TITLE
Coalesce an IME-composed character into one undo unit

### LINKED ISSUE #318: Undo after Korean (IME) input toggles character instead of removing it
### What happened:

In the document editor, undo behaves incorrectly after typing Korean (Hangul) characters via IME. A single typed Hangul character appears to be registered as **two separate undo history entries**, so undo seems to "toggle" the character between visible and hidden states before it is actually removed.

Concretely, after typing a single Hangul character:

- 1st Undo: the character disappears
- 2nd Undo: the same character reappears
- 3rd Undo: it disappears again
- 4th Undo: it reappears again

The character is only truly removed from the document after an even number of undos. Odd-numbered undos restore it. Both isolated vowels (e.g. "ㅏ") and fully composed Hangul syllables (e.g. "가") show this pattern.

English input does not exhibit this issue — undo works normally character by character.

This strongly suggests the undo history is pushing both the IME `compositionstart` (or interim composition state) and `compositionend` as separate history entries, instead of merging them into a single user-visible insertion.

### What you expected to happen:

A single Hangul character typed via IME should be a single undo history entry. One Undo should remove it cleanly, matching the behavior of typing a single English character.

This matches the behavior of Google Docs, Notion, and other collaborative editors that handle IME composition.

### How to reproduce it (as minimally and precisely as possible):

1. Open a document on https://wafflebase.io/
2. Switch the system input source to Korean (Hangul) IME
3. Click into a regular paragraph (not a table cell)
4. Type a single Hangul character, e.g. "가" (composed of ㄱ + ㅏ) or just "ㅏ"
5. Press Cmd + Z once → the character disappears
6. Press Cmd + Z again → the character reappears
7. Repeat → the character toggles between visible and hidden

For comparison:

1. Switch the input source to English
2. Type a single English character, e.g. "a"
3. Press Cmd + Z once → the character is cleanly removed (no toggling)

### Anything else we need to know?:

- This is specific to Korean IME input. English input works correctly, which points to composition-event handling rather than the undo stack itself being broken.
- The most likely cause is the undo history capturing IME composition lifecycle events (`compositionstart` / `compositionupdate` / `compositionend`) as multiple entries instead of collapsing them into a single insertion. Other CJK IMEs (Japanese, Chinese) likely exhibit the same behavior — worth testing as a control.
- Occurs in plain paragraphs. I haven't isolated whether it also occurs in tables or headings; the behavior is consistent enough in plain paragraphs to file separately.
- Filing this for visibility — I'm not planning to take this on myself right now.

### Environment:

- Operating system: macOS Sonoma 14.6.1
- Browser and version: Chrome 148.0.7778.96 (arm64)
- Input method: Korean — macOS 2-Set Korean (system default)

### PR DESCRIPTION
**Pull request by semimikoh**
## Summary

Fixes the Hangul IME undo "toggle" (issue #318): typing one Korean syllable used
to land as **multiple overlapping Yorkie undo units**, so Undo flipped the
character on/off instead of removing it once. Now **one IME-composed character =
one undo unit**, matching English input and Google Docs / Notion.

Opened design-note-first; now contains the full implementation on the same
branch.

## Why

Undo/redo is Yorkie's document history: each `doc.update()` is exactly one undo
unit, and (confirmed against `@yorkie-js/sdk` 0.7.8) there is no way to exclude a
Tree edit from history or group across updates. The IME pipeline wrote interim
composition into the Tree on every `compositionupdate` (delete + insert) and
again at `compositionend`, so one syllable produced 3–5 reversible edits at the
same position — the toggle.

**Approach B (view-local layout injection):** interim composing text is never
written to the model; it is injected into the *layout* as a synthetic inline at
the caret, so it wraps/reflows like real text while the model stays untouched.
The character is committed exactly once on `compositionend` → one `doc.update()`
→ one undo unit. (Approach A — painting at the caret — was rejected: it doesn't
reflow, so mid-paragraph composition would overlap following text.)

Body / header / footer / table-cell all funnel through the same `layoutBlock`,
so one seam covers every region.

## Linked Issues

Fixes #318

## Author checklist

- [x] I searched existing issues and PRs and confirmed this is not a duplicate.
- [x] I read every line of this diff and can explain why each change is necessary.
- [x] Changes follow the conventions in the touched packages; any deviations are explained in *Why* above.
- [x] If AI tools assisted with this PR, I noted where in *Notes for Reviewers* below.

## Verification

CI automatically posts a verification summary comment on this PR with
per-lane results for both `verify:self` and `verify:integration`.

- [x] verify:self — green locally (pre-push lanes: lint, unit, builds, chunk budgets, entropy); CI to re-confirm.
- [x] verify:integration — CI comment shows ✅ (or explicit skip reason below)

Skip reason (if applicable): the only Yorkie-path change is a **frontend
presence-offset clamp** (`clampPosToModel` in `yorkie-doc-store.ts`); no
backend/Yorkie-server logic changed. Covered by unit tests against a real
`yorkie.Document`. CI `verify:integration` runs automatically regardless.

## Risk Assessment

- User-facing risk: touches the Docs text-input/composition path. Mitigated by
  23 new unit tests (layout injection + reflow, editor IME behavior, undo units,
  presence clamp) and manual smoke (one-step undo, mid-paragraph reflow, blur,
  software-Hangul) in the standalone playground and the full Yorkie app.
- Data/security risk: none. No schema/auth/persistence change; interim composing
  text is never written to the shared model.
- Rollback plan: revert this PR's commits — isolated to docs layout/text-editor
  + one frontend presence clamp; no migrations or API changes.

## Notes for Reviewers

- UI changes: no new chrome; behavior change is IME composition rendering + undo
  granularity. (Composing text rendering is unchanged visually; it just no longer
  hits the model.)
- Incorporates two rounds of self-review: presence-leak clamp, stale-preview
  re-emit on remote anchor correction, blur ghost-text cleanup + `compositionEnd`
  lifecycle, and caret-follows-composing-text on remote correction.
- AI tools assisted (codebase/SDK investigation, design note, implementation)
  using Claude Code (Claude Opus 4.8); the author reviewed the root-cause
  analysis and every change.
- Follow-up: the docs playground IME smoke script (`test:ime`) pre-dates this
  work and is broken against the Alice-seeded demo (expects an empty doc) —
  tracked separately, not addressed here.

### CODE REVIEW

**Comment by hackerwins** [comment]
Direction looks good. Feel free to push the implementation on this branch.

File: packages/docs/src/view/text-editor.ts [hunk 1]
```diff
@@ -227,7 +241,26 @@ export class TextEditor {
...
+    this.cursor.moveTo(caretPos, this.getWrapAffinity(caretPos));
+    // Re-publish the view-local composing injection at the corrected
+    // position so the preview does not render at the stale offset until the
+    // next input event. Mark both the old and new block dirty so the old
+    // block sheds its injection and the new one picks it up.
+    this.markDirty(prevBlockId);
+    this.markDirty(pos.blockId);
+    this.emitComposingContext();
+    this.requestRender();
   }
```
[the remark this answers is not in the document]
Comment (semimikoh) [reply packages/docs/src/view/text-editor.ts:261]: @coderabbitai Acknowledged — this is a real gap, but I'm deferring it as a follow-up rather than addressing it in this PR. Reasoning:

1. The remote drift-correction lifecycle (`isComposing()` / `setCompositionStartPosition()` / the composition listeners) has always been browser-IME-only — it was introduced for the browser path in #257 and this PR didn't add that boundary.
2. This PR's scope is the issue #318 undo-toggle fix. The software-Hangul path is the Mobile Safari fallback, and the failing scenario requires fallback composition concurrent with a remote edit *mid-composition* — a narrow edge case.
3. Folding both paths behind a single composition-state object is a sizable refactor that deserves its own change with dedicated remote-edit tests, rather than being bundled into the undo fix.

Tracking as a follow-up; not blocking this PR.

File: packages/frontend/src/app/docs/yorkie-doc-store.ts [hunk 2]
```diff
@@ -863,6 +863,33 @@ export class YorkieDocStore implements DocStore {
...
+      const currentDoc = this.getDocument();
+      const { path, region } = this.resolveBlockTreePath(pos.blockId, currentDoc);
+      const block = this.getBlockByRegion(currentDoc, path, region);
+      if (!block) return pos;
+      const len = this.blockTextLength(block);
+      return { blockId: pos.blockId, offset: Math.max(0, Math.min(pos.offset, len)) };
+    } catch {
+      return pos;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (semimikoh) [reply packages/frontend/src/app/docs/yorkie-doc-store.ts:891]: @coderabbitai Fair point that the `try/catch` can read as graceful degradation while the next line still throws. But the stale-block-id throw is **pre-existing**, not introduced here: before this PR, `updateCursorPos()` already called `anchorDocPosition(pos)` directly with the raw position, which throws the same way when the block was deleted by a remote peer.

`clampPosToModel()` was added narrowly to clamp an out-of-model *offset* during IME composition (the concern this PR targets) — not to make the whole presence-publish path resilient to deleted-block cursors. Hardening `updateCursorPos()` / `anchorDocPosition()` against stale block ids needs its own remote-delete test scenarios and is out of scope for the #318 fix, so I'm deferring it as a follow-up.

**Comment by hackerwins** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
